File: go/oop-patterns/creational-patterns/multiton/multiton.go
package main

import "fmt"

type singletonType int

const (
	нулевой singletonType = iota
	первый
	второй
	третий
)

type singleton struct {
	data singletonType
}

func (s *singleton) toString() string {
	return fmt.Sprintf("Я одиночка типа %v", s.data)
}

func newSingleton(t singletonType) *singleton {
	return &singleton{data: t}
}

// Мультитон содержит список ссылок на объекты-одиночки
// разных типов и возвращает ссылку на объект конкретного типа
// в ответ на запрос клиента, либо создаёт объект конкретного
// типа, сохраняет его в своём списке и возвращает ссылку на
// него, если одиночки такого типа ещё нет в списке на
// момент запроса
type multiton struct {
	instances map[singletonType]*singleton
}

func (m *multiton) getInstance(t singletonType) *singleton {
	if _, ok := m.instances[t]; !ok {
		m.instances[t] = newSingleton(t)
	}
	return m.instances[t]
}

func newMultiton() *multiton {
	return &multiton{instances: make(map[singletonType]*singleton)}
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func main() {
	m := newMultiton()
	fmt.Println(m.getInstance(первый).toString())
	fmt.Println(m.getInstance(второй).toString())
	fmt.Println(m.getInstance(третий).toString())
}

File: javascript/oop-patterns/creational-patterns/multiton/multiton.js
const SingletonType = {
  Первый: 1,
  Второй: 2,
  Третий: 3,
};

// Мультитон содержит список ссылок на объекты-одиночки
// разных типов и возвращает ссылку на объект конкретного типа
// в ответ на запрос клиента, либо создаёт объект конкретного
// типа, сохраняет его в своём списке и возвращает ссылку на
// него, если одиночки такого типа ещё нет в списке на
// момент запроса
class Multiton {
  static instances = new Map();

  static getInstance(type) {
    if (!this.instances.has(type)) {
      this.instances.set(type, new Single(type));
    }
    return this.instances.get(type);
  }
}

class Single {
  data;
  constructor(type) {
    this.data = type;
  }
  toString() {
    return `Я одиночка типа ${this.data}`;
  }
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

console.log(Multiton.getInstance(SingletonType.Первый).toString());
console.log(Multiton.getInstance(SingletonType.Второй).toString());
console.log(Multiton.getInstance(SingletonType.Третий).toString());
